bundler/lib/bundler/lockfile_generator.rb

```diff
@@ -65,6 +66,50 @@ def add_dependencies
       end
     end
 
+    def add_checksums
+      out << "\nCHECKSUMS\n"
+
+      definition.resolve.sort_by(&:full_name).each do |spec|
+        if spec.respond_to?(:to_checksum)
+          if spec.is_a?(LazySpecification) && spec.generic_platform_with_specific_installed_platform?
+            # If a spec started off as a generic ruby platform, and is resolved
+            # to something more specific on install time, we skip recording the
+            # checksum.
+            # For example if we have:
+            #
+            #     specs:
+            #       pry (0.11.3)
+            #
+            #   PLATFORMS
+            #     ruby
+            #     java
+            #
+            # Recording the checksum for ruby does not make sense, because we
+            # will end up with something like:
+            #
+            #     specs:
+            #       pry (0.11.3)
+            #
+            #   CHECKSUMS
+            #     pry(0.11.3-java)
+            #       1f0bd9ce0282e8a20151011e4677746206a0c84b648e0a1c981b3232bb47a3b9
+            #     pry(0.11.3-java)
+            #       1f0bd9ce0282e8a20151011e4677746206a0c84b648e0a1c981b3232bb47a3b9
+            #
+            #   PLATFORMS
+            #     ruby
+            #     java
+            next
```

Comment:
Due to https://github.com/rubygems/rubygems/issues/3379#issuecomment-1196299629, I had to skip "generic" platforms like "ruby". 

I think this means gems like `nokogiri` will not have their checksums locked under legacy lockfiles where `ruby` is the only `PLATFORMS.

Reply:
From the list of compact entries, here is a non-exhaustive list of gems that have more than one platform

```
index = Bundler::Fetcher::CompactIndex.new(nil, Bundler::Source::Rubygems::Remote.new(Bundler::URI("https://rubygems.org")), nil)
dependencies = Dir[index.send(:compact_index_client).instance_variable_get(:@cache).directory.join("info").join('*')].map {|d| File.basename d}
dependencies.select do |d|
  index.send(:compact_index_client).instance_variable_get(:@cache).dependencies(d).group_by(&:first).values.last.size > 1
end
```

```
["nio4r", "aws-crt", "msgpack", "mixlib-shellout", "ffi-yajl", "concurrent-ruby-ext", "strscan", "mysql2", "sorted_set", "puma", "rest-client", "io-wait", "hpricot", "io-console", "libxml-ruby", "unf_ext", "grpc", "win32console", "google-protobuf", "ruby-prof", "bcrypt", "chef", "hamlit", "digest", "date", "websocket-driver", "hiredis", "bcrypt_pbkdf", "binding_ninja", "murmurhash3", "ptools", "atomic", "eventmachine", "mongrel", "pg", "psych", "cgi", "bcrypt-ruby", "win32-api", "sassc", "json", "stringio", "snappy", "mixlib-archive", "ffi", "sqlite3", "racc", "pry", "libv8", "thread_safe", "jaro_winkler", "character_set", "sys-admin", "ed25519", "nokogiri", "jruby-pageant", "regexp_property_values", "linecache"]
```